test/gtest/ucp/test_ucp_request.cc

```diff
@@ -580,46 +566,54 @@ UCP_INSTANTIATE_TEST_CASE(test_proto_reset)
 #if ENABLE_DEBUG_DATA
 class test_proto_reset_rndv_get : public test_proto_reset {
 protected:
-    void wait_and_restart(const std::vector<request_pair_t> &pairs) override
+    void wait_and_restart(const std::vector<void*> &reqs) override
     {
-        wait_for_condition(pairs, [](const request_pair_t &pair) {
-            if (pair.rreq->recv.proto_rndv_request == NULL) {
+        wait_any(reqs, [](const void *ureq) {
+            if (ureq == NULL) {
+                return false;
+            }
+
+            ucp_request_t *req = (ucp_request_t*)ureq - 1;
+            if (req->flags & UCP_REQUEST_FLAG_PROTO_SEND) {
                 return false;
             }
 
-            const ucp_request_t *rndv_req = pair.rreq->recv.proto_rndv_request;
+            const ucp_request_t *rndv_req = req->recv.proto_rndv_request;
 
-            return (rndv_req->send.state.dt_iter.offset > 0) &&
-                   (rndv_req->send.state.dt_iter.offset <
-                    rndv_req->send.state.dt_iter.length);
+            if (rndv_req == NULL) {
+                return false;
+            }
+
+            const ucp_datatype_iter_t *dt_iter = &rndv_req->send.state.dt_iter;
+            return !ucp_datatype_iter_is_begin(dt_iter) &&
+                   !ucp_datatype_iter_is_end(dt_iter);
```

Comment:
can we share some common code with line 341-355?

Reply:
In 341 I test send request and in 583 recv request. which means:
1) check PROTO_SEND exist for send request, but not exist for recv request.
2) data_iter for send request is done on the request itself, in contrast with recv request which check internal RNDV request.

Reply:
some of the code looks similar so could use a helper function